pydra/engine/task.py

```diff
@@ -0,0 +1,443 @@
+# -*- coding: utf-8 -*-
+"""task.ipynb
+
+Automatically generated by Colaboratory.
+
+Original file is located at
+    https://colab.research.google.com/drive/1RRV1gHbGJs49qQB1q1d5tQEycVRtuhw6
+
+## Notes:
+
+### Environment specs
+1. neurodocker json
+2. singularity file+hash
+3. docker hash
+4. conda env
+5. niceman config
+6. environment variables
+
+### Monitors/Audit
+1. internal monitor
+2. external monitor
+3. callbacks
+
+### Resuming
+1. internal tracking
+2. external tracking (DMTCP)
+
+### Provenance
+1. Local fragments
+2. Remote server
+
+### Isolation
+1. Working directory
+2. File (copy to local on write)
+3. read only file system
+"""
+
+
+import abc
+import dataclasses as dc
+import datetime as dt
+from hashlib import sha256
+import json
+import os
+import pickle as cp
+from tempfile import mkdtemp
+import typing as ty
+import inspect
+import asyncio
+from pathlib import Path
+
+File = ty.NewType('File', Path)
+Directory = ty.NewType('Directory', Path)
+
+
+def ensure_list(obj):
+    if obj is None:
+        return []
+    if isinstance(obj, list):
+        return obj
+    return [obj]
+
+
+def print_help(obj):
+    helpstr = 'help {}'.format(obj)
+    print(helpstr)
+    return helpstr
+
+
+def load_result(checksum, cache_locations):
+    if not cache_locations:
+        return None
+      
+    for location in cache_locations:
+        if (location / checksum).exists():
+            return cp.loads(
+                (location / checksum / '_result.pklz').read_bytes())
+    return None
+
+
+def save_result(result_path: Path, result):
+    with (result_path / '_result.pklz').open('wb') as fp:
+        return cp.dump(dc.asdict(result), fp)
+
+
+def task_hash(task_obj):
+    """
+    input hash, output hash, environment hash
+    
+    :param task_obj: 
+    :return: 
+    """
+    return NotImplementedError
+
+
+def now():
+    return dt.datetime.utcnow().isoformat(timespec='microseconds')
+
+
+prov_context = "https://openprovenance.org/prov.jsonld"
+schema_context = "https://schema.org/docs/jsonldcontext.json"
+pydra_context = {"pydra": "https://uuid.pydra.org/"}
+
+# audit flags
+AUDIT_PROV = 0x01
+AUDIT_RESOURCE = 0x02
```

Comment:
If we're 3.6+, we can use [`enum.Flag`](https://docs.python.org/3/library/enum.html#flag).

```Python
class AuditFlag(enum.Flag):
    PROV = enum.auto()  # 0x01
    RESOURCE = enum.auto()  # 0x02
```

Reply:
implemented.